public/tier1/keyvalues3.h

```diff
@@ -0,0 +1,1027 @@
+#ifndef KEYVALUES3_H
+#define KEYVALUES3_H
+
+#ifdef _WIN32
+#pragma once
+#endif
+
+#include "tier0/platform.h"
+#include "tier0/dbg.h"
+#include "tier1/bufferstring.h"
+#include "tier1/generichash.h"
+#include "tier1/strtools.h"
+#include "tier1/utlbuffer.h"
+#include "tier1/utlhashtable.h"
+#include "tier1/utlleanvector.h"
+#include "tier1/utlmap.h"
+#include "tier1/utlstring.h"
+#include "tier1/utlsymbollarge.h"
+#include "mathlib/vector4d.h"
+#include "Color.h"
+#include "entityhandle.h"
+
+#include "tier0/memdbgon.h"
+
+class KeyValues3;
+class CKeyValues3Array;
+class CKeyValues3Table;
+class CKeyValues3Cluster;
+class CKeyValues3ArrayCluster;
+class CKeyValues3TableCluster;
+class CKeyValues3Context;
+struct KV1ToKV3Translation_t;
+struct KV3ToKV1Translation_t;
+
+/* 
+	KeyValues3 is a data storage format. See https://developer.valvesoftware.com/wiki/KeyValues3
+	Supports various specific data types targeted at the Source2.
+	Each specific type corresponds to one of the basic types.
+
+	There are 2 ways to create KeyValues3:
+
+	1. Via CKeyValues3Context:
+	- KV's, arrays and tables are stored in fixed memory blocks (clusters) and therefore memory is allocated only when clusters are created.
+	- Supports metadata and some other things.
+
+	2. Directly through the constructor.
+*/
+
+struct KV3ID_t
+{
+	const char* m_name;
+	uint64		m_data1;
+	uint64		m_data2;
+};
+
+const KV3ID_t g_KV3Encoding_Text = 
+{
+	"text",
+	0x41C58A33E21C7F3Cull,
+	0xDAA323A6DA77799ull
+};
+
+const KV3ID_t g_KV3Format_Generic = 
+{
+	"generic",
+	0x469806E97412167Cull,
+	0xE73790B53EE6F2AFull
+};
+
+enum KV1TextEscapeBehavior_t
+{
+	KV1TEXT_ESC_BEHAVIOR_UNK1 = 0,
+	KV1TEXT_ESC_BEHAVIOR_UNK2 = 1,
+};
+
+enum KV3SaveTextFlags_t
+{
+	KV3_SAVE_TEXT_TAGGED = (1 << 0), // adds subtype name before value
```

Comment:
This potentially could have a ``KV3_SAVE_TEXT_NONE`` and be used in the SaveKV3* functions below as a default ``flags`` argument value